### ISSUE TITLE
Investigate if ECJ's parser be simplified by incorporating overlooked last minute spec revisions 

### ISSUE THREAD

**Issue by srikanth-sankaran** [eclipse-jdt/eclipse.jdt.core#1045]
Per observations from @stephan-herrmann here: https://github.com/eclipse-jdt/eclipse.jdt.core/issues/125#issuecomment-1489366453
we may have inadvertently overlooked last minute changes to Java 8 spec that may have allowed us to simplify the Parser/Scanner in ECJ.

See 

https://bugs.eclipse.org/bugs/show_bug.cgi?id=561934#c4
https://bugs.eclipse.org/bugs/show_bug.cgi?id=561934#c10

**Comment by trancexpress** [comment]
Thanks for looking into this @srikanth-sankaran !

**Comment by stephan-herrmann** [comment]
In an attempt to clarify what can/cannot be done I simplified the Java grammar to a minimum that still shows a remaining conflict:

[mini2.txt](https://github.com/eclipse-jdt/eclipse.jdt.core/files/13697841/mini2.txt)
Rename this file to `mini.g` and pass it to `jikespg`, it will show the dreaded
```
*** Shift/reduce conflict on "MULTIPLY" with rule 2

This grammar is not LALR(1).
```

In all the rest of the Java grammar, productions are chained in a way that anything on the RHS always has a reduced set of possibilities, ensuring that the right-most non-terminal will never branch back to the current LHS.

In the reduced version `Cast ::= LParent Identifier RParent Lambda` violates that rule, because `Lambda` may end with a generic `Expr`.

As a result this expression is ambiguous:
```
(Type) () -> a * b
```

The parser cannot decide which way to interpret (denoted by resulting AST):
* Mul ( Cast ( Lambda (Identifier)), Identifier )
* Cast ( Lambda ( Mul (Identifier, Identifier) ) )

IOW, this mini grammar (and transitively the Java grammar) cannot decide if the trailing `*b` binds to the inner expression `a` or to the entire expression so far `(Type) () -> a`.

I couldn't yet find documentation of all of those options of jikespg, if something there allows to define precedence to the "inner" rule?

Semantically, multiplying a cast with some factor _normally_ doesn't make sense, but when casting to a numerical boxing type this may still do something.

Casting a lambda, however, will never produce a numerical type, as none of those are functional interfaces, not sure if this can be turned into something useful at the syntax level ...

**Comment by stephan-herrmann** [comment]
The following change would fix the mini grammar, to be LALR(1):
```
--- mini-nok.g
+++ mini-fixed.g
@@ -37,6 +37,7 @@
 
 Expr -> Mul
 Expr -> Lambda
+Expr -> CastL
 
 Mul -> Unary
 Mul ::= Mul MULTIPLY Unary
@@ -45,7 +46,7 @@
 Unary -> Cast
 
 Cast ::= LParen Identifier RParen Primary
-Cast ::= LParen Identifier RParen Lambda
+CastL ::= LParen Identifier RParen Lambda
 
 Primary -> Identifier
```

**Explanation:** I pulled the casted-lambda rule out of the entire tree of productions applying some kind of arithmetic, assuming that you simply cannot "compose" a (casted) lambda with any other value.

Would this illegally enable use of casted-lambda in a context where it's illegal? I think that could be compensated by either of
* find a more suitable nonterminal to hook into
* detect illegal use later during compilation

Indeed also the actual `java.g` can be disambiguated using this trick :tada: :smile: :sparkles: 

_PS: I could imagine that such change would even be worth pushing upstream, i.e., to JLS._

**Comment by stephan-herrmann** [comment]
Java is funny.

A casted lambda expression can actually be used as an operand in certain expressions. 

Need to check if we break anything regarding:
* instanceof
* != and == comparisons
* ternaries
* assignment expressions

E.g.:
```java
  	if (((Supplier<String>) () -> a) instanceof Supplier)
	    System.out.print("yes");
  	if (((Supplier<String>) () -> a) != null)
	    System.out.print("yes");
```
This is accepted by javac and prints "yesyes" :)

From experiments in this area I learned that `instanceof` normally binds to the inner expression, to bind it to the lambda parens are needed.

Actually, ecj barfs on that:
```
2. ERROR in /tmp/X.java (at line 13)
        if (((Supplier<String>) () -> "a") != null)
                                ^^^^^^^^^
Missing code implementation in the compiler
```
Let's hope that grammar simplification will avoid this unspeakable error message, and causes no further havoc. In fact I wonder if anyone would insist that ecj supports checking a lambda for null, or performing instanceof checks. In both cases the lambda will be left to gc before it can ever be invoked.

**Comment by srikanth-sankaran** [comment]
> ```java
> if (((Supplier<String>) () -> a) instanceof Supplier)
> 	    System.out.print("yes");
>   	if (((Supplier<String>) () -> a) != null)
> 	    System.out.print("yes");
> ```

I have raised https://github.com/eclipse-jdt/eclipse.jdt.core/issues/1755

**Comment by srikanth-sankaran** [comment]
Thanks for the various experiments and the suggestions Stephan, I will study these. I am missing my Dragon book today, will have it in my hands tomorrow.

**Comment by stephan-herrmann** [comment]
Genie reminded me that we have some unfinished work still in gerrit:
* https://bugs.eclipse.org/bugs/show_bug.cgi?id=573152
* https://git.eclipse.org/r/c/jdt/eclipse.jdt.core/+/179810

Seeing mention of `fallBackToSpringForward()` and friends, I guess we want to wait for the grammar simplification, before polishing this code area. Perhaps some of the magic will no longer be needed at some point :)

**Comment by srikanth-sankaran** [comment]
> In an attempt to clarify what can/cannot be done I simplified the Java grammar to a minimum that still shows a remaining conflict:
> 
> [mini2.txt](https://github.com/eclipse-jdt/eclipse.jdt.core/files/13697841/mini2.txt) Rename this file to `mini.g` and pass it to `jikespg`, it will show the dreaded
> 
> ```
> *** Shift/reduce conflict on "MULTIPLY" with rule 2
> 
> This grammar is not LALR(1).
> ```
> 
> In all the rest of the Java grammar, productions are chained in a way that anything on the RHS always has a reduced set of possibilities, ensuring that the right-most non-terminal will never branch back to the current LHS.
> 
> In the reduced version `Cast ::= LParent Identifier RParent Lambda` violates that rule, because `Lambda` may end with a generic `Expr`.
> 
> As a result this expression is ambiguous:
> 
> ```
> (Type) () -> a * b
> ```
> 
> The parser cannot decide which way to interpret (denoted by resulting AST):
> 
> * Mul ( Cast ( Lambda (Identifier)), Identifier )
> * Cast ( Lambda ( Mul (Identifier, Identifier) ) )
> 
> IOW, this mini grammar (and transitively the Java grammar) cannot decide if the trailing `*b` binds to the inner expression `a` or to the entire expression so far `(Type) () -> a`.
> 
> I couldn't yet find documentation of all of those options of jikespg, if something there allows to define precedence to the "inner" rule?

I don't know either - but I am not averse to, (at some point) - if such an options are not available, building them on an experimental basis to see where that would lead us. IIRC, YACC allows you to say precedence between two rules for shift/reduce (and where no precedence is specified always prefers shift) and for reduce/reduce it always opt for the earlier listed rule.  

> 
> Semantically, multiplying a cast with some factor _normally_ doesn't make sense, but when casting to a numerical boxing type this may still do something.
> 
> Casting a lambda, however, will never produce a numerical type, as none of those are functional interfaces, not sure if this can be turned into something useful at the syntax level ...

**Comment by srikanth-sankaran** [comment]
Two big blockers for certain types of experiments in the area of parser/scanner are:

1. The DiagnoseParser with its too much voodoo in the works. The fact that there are multiple machines that are run in parallel with too much of the automaton's innards and tables being manipulated and more importantly, the fact that tokens are read ahead of time and buffered into a stream mean that we can never have a satisfactory implementation of `org.eclipse.jdt.internal.compiler.parser.diagnose.DiagnoseParser.automatonWillShift(int)` (that can be useful for the scanner)

This method is a fantastic side-effect free mechanism to query the parser of its current state - It is axiomatic that an LALR parser will never shift on invalid input - so this is a great tool to steer the scan and the parse.

Presently this method is used to disambiguate when `when` is a contextual keyword vs a plain old identifier. Ask the parser if it will shift the keyword `when` and if it answers yes, then the when in the input stream is a contextual keyword and otherwise otherwise.

It is possible to build this disambiguation into DiagnoseParser's internal token stream but that is some work.

2. CompletionParser which works with incomplete evolving, broken code.

In general building smarts falls in some buckets. 

(a) Feedback from the parser to the scanner: this allows consumed input to steer classification of unconsumed input. This is looking back in the rear view mirror. (b) Looking ahead to detect structure of what is to come to classify current token (perhaps inject some synthetics to steer the parser in a particular path) - this is Vanguard parser's forte.
and 
(c) The now abandoned String Templates work had a brand new way of dealing conflicts in the grammar - by making some non-terminals symbols (with their own grammatical structure) appear as (pseudo) terminal symbols. Unfortunately before we could gain good hands on experience with that model, we had to rip it out.

**Comment by stephan-herrmann** [comment]
> Two big blockers for certain types of experiments in the area of parser/scanner are:

I see that there are interesting tasks still ahead of us, but does any of what you mention here block the current grammar refactoring?

**Comment by stephan-herrmann** [comment]
> > I couldn't yet find documentation of all of those options of jikespg, if something there allows to define precedence to the "inner" rule?
> 
> I don't know either - but I am not averse to, (at some point) - if such an options are not available, building them on an experimental basis to see where that would lead us. IIRC, YACC allows you to say precedence between two rules for shift/reduce (and where no precedence is specified always prefers shift) and for reduce/reduce it always opt for the earlier listed rule.

I spent a little while on jikespg sources and tried if any options would have the desired effect, but to no avail.
The most promising option name was SHIFT-DEFAULT, but this flag seems only to enable an optimization, whereby for each terminal a default shift is determined, such that this default can be eliminated from the state automaton. I guessed this from these code comments:
```
/****************************************************************************/
/*                            COMPUTE_SHIFT_DEFAULT:                        */
/****************************************************************************/
/* This procedure updates the vector SHIFTDF, indexable by the terminals in */
/* the grammar. Its task is to assign to each element of SHIFTDF, the action*/
/* most frequently defined on the symbol in question.                       */
/****************************************************************************/
```
and
```
/*****************************************************************************/
/*                             COMPUTE_GOTO_DEFAULT:                         */
/*****************************************************************************/
/*   COMPUTE_GOTO_DEFAULT constructs the vector GOTODEF, which is indexed by */
/* the non-terminals in the grammar. Its task is to assign to each element   */
/* of the array the Action which is most frequently defined on the symbol in */
/* question, and remove all such actions from the state automaton.           */
/*****************************************************************************/
```

And then my C-reading skills are not sufficient to figure out where such an option could be added on our own. Or perhaps it's my lack of understanding of what a parser generator actually does :)

**Comment by srikanth-sankaran** [comment]
> > Two big blockers for certain types of experiments in the area of parser/scanner are:
> 
> I see that there are interesting tasks still ahead of us, but does any of what you mention here block the current grammar refactoring?

No, they don't block any work that eliminates conflicts in the first place by rewriting the grammar. I was jotting down the thoughts about what our options are for handling a conflict.

**Comment by srikanth-sankaran** [comment]
@stephan-herrmann @mpalat @jarthana - Can use some feedback on this following question:

I see that handshake between the Parser and Scanner has been steadily growing. The comment above `org.eclipse.jdt.internal.compiler.parser.Scanner.VanguardParser.parse(Goal)` clearly says: "Canonical LALR pushdown automaton identical to Parser.parse() **minus side effects of any kind**" but many consume* methods of the parser have started introducing side effects to clue in the scanner about where they are in the parse.

See for example, the state held by 

`org.eclipse.jdt.internal.compiler.parser.Scanner.caseStartPosition`,
`org.eclipse.jdt.internal.compiler.parser.Scanner.yieldColons`
`org.eclipse.jdt.internal.compiler.parser.Scanner.multiCaseLabelComma`

Now there are two problems with this: one is that `VanguardParser` doesn't execute the consume* methods so the side effects will be missing. Secondly DiagnoseParser won't execute the consume* methods so the handshake between the Scanner and Parser will be absent in DiagnoseParse which means in the presence of syntax errors, the scanner will likely be clueless while tokenizing some advanced constructs that need cooperation from parser and scanner.

The comment in `org.eclipse.jdt.internal.compiler.parser.Scanner.VanguardScanner.getNextToken()` is interesting:

"// this.caseStartPosition > this.startPositionpossible on recovery - **bother only about correct ones.**"

But if that is the approach we want to take - which may not be unreasonable - there is no need for so many state driven hacky handshakes.

`org.eclipse.jdt.internal.compiler.parser.ConflictedParser.automatonWillShift(int)` may help us address many of situations without using Vanguard parser : see https://github.com/eclipse-jdt/eclipse.jdt.core/issues/456 for one instance of what I mean

**Comment by srikanth-sankaran** [comment]
> @stephan-herrmann @mpalat @jarthana - Can use some feedback on this following question:

The question I am asking is: Given that the state based handshake between the Parser and the Scanner won't work in DiagnoseParse and VanguardParse, why not ditch that and opt for a cleaner solution in disambiguating token by using  `org.eclipse.jdt.internal.compiler.parser.ConflictedParser.automatonWillShift(int)` where possible which approach would also not work with the DiagnoseParse (without heroic effort ... actually scratch that - without super-heroic effort) but which may render a VanguardParse unnecessary in that scenario. 

See that `org.eclipse.jdt.internal.compiler.parser.Scanner.disambiguatedRestrictedIdentifierWhen(int)` does not run the VanguardParser to disambiguate `when` - 

(See also `org.eclipse.jdt.internal.compiler.parser.diagnose.DiagnoseParser.automatonWillShift(int)` which is clueless how to answer the question - the comment in the earlier method `org.eclipse.jdt.internal.compiler.parser.diagnose.DiagnoseParser.atConflictScenario(int)`) is applicable here too)

**Comment by stephan-herrmann** [comment]
@srikanth-sankaran  can we, pretty please, try to exercise some separation of concerns where it is still possible?

This issue is about modifying the grammar to the end that hopefully less disambiguation / lookahead / scanner-parser communication is necessary.

If you want to discuss how to improve the remaining kludges for handling conditional keywords and grammar ambiguities, please create a separate space for that. Maybe add a paragraph or two in the structure below https://github.com/eclipse-jdt/eclipse.jdt.core/wiki/ECJ ? Have you seen https://github.com/eclipse-jdt/eclipse.jdt.core/wiki/ECJ-Parse ?

**Comment by srikanth-sankaran** [comment]
> I am missing my Dragon book today, will have it in my hands tomorrow.

Sorry that is embarrassing 😜 

Clear statement of priority: Any regressions I injected in the last few months that get reported get highest priority. Second highest priority is for the present topic and [https://github.com/eclipse-jdt/eclipse.jdt.core/issues/1045

**Comment by srikanth-sankaran** [comment]
> From experiments in this area I learned that `instanceof` normally binds to the inner expression, to bind it to the lambda parens are needed. This could mean that these rules will never directly branch into (casted) lambda but need `( Expression )` to get there, in which case the parens will disambiguate.

I think this is true of expression bodied lambdas. The following should continue to compile with any modified grammar:

```
import java.util.function.Supplier;

public class X {
    public static void main(String[] args) {
		boolean b = (Supplier<String>) () -> { return "Hello"; } instanceof Supplier<String>;
	}
}
```

**Comment by srikanth-sankaran** [comment]
> The following should continue to compile with any modified grammar:

I mean ideally we should continue to ...

> In fact I wonder if anyone would insist that ecj supports checking a lambda for null, or performing instanceof checks. In both cases the lambda will be left to gc before it can ever be invoked.

I share the skepticism. If it means bending over backwards to accommodate it, I would be inclined to say don't bother. This is extreme corner case.

**Comment by srikanth-sankaran** [comment]
> Need to check if we break anything regarding:
> 
> * instanceof
> * != and == comparisons
> * ternaries
> * assignment expressions

This piece of code would break if we hook a `CastedLambdaExpression` to a top level `Expression`

```
import java.util.function.Supplier;

public class X {
    public static void main(String[] args) {
    	Supplier<String> s = () -> "Hello";
		boolean b = true;
		Supplier<String> sup = b ? s : (Supplier<String>) () -> { return "World"; };
	}
}
```

Also this would break (different rule reductions from the above):

```
import java.util.function.Supplier;

public class X {
	public static void main(String[] args) {
    	Supplier<String> s = () -> "Hello";
		boolean a = true, b = false;
		Supplier<String> sup  = (a ? s : (Supplier<String>) () -> { return "World"; });	
	}
}
```

The following breaks as is with the PR https://github.com/eclipse-jdt/eclipse.jdt.core/pull/4667 although it doesn't have any casts:

```
import java.util.function.Supplier;

public class X {
	public static void main(String[] args) {
    	Supplier<String> s = () -> "Hello";
		boolean a = true, b = false;
		Supplier<String> sup  = (a ? s : () -> { return "World"; });	
	}
}
```

**Comment by srikanth-sankaran** [comment]
@stephan-herrmann - what explains the difference in these productions relative to every other xxx_NotName  productions ? Any idea ? I am asking what explains the use of `ShiftExpression_NotName` in the RHS of these rules (instead of `RelationalExpression_NotName `) ? While every other xxx_NotName seems to use the LHS non terminal as the first operator of the binary op.

RelationalExpression_NotName -> ShiftExpression_NotName
RelationalExpression_NotName ::= **ShiftExpression_NotName** '<' ShiftExpression
/.$putCase consumeBinaryExpression(OperatorIds.LESS); $break ./
RelationalExpression_NotName ::= Name '<' ShiftExpression
/.$putCase consumeBinaryExpressionWithName(OperatorIds.LESS); $break ./
RelationalExpression_NotName ::= **ShiftExpression_NotName** '>' ShiftExpression
/.$putCase consumeBinaryExpression(OperatorIds.GREATER); $break ./
RelationalExpression_NotName ::= Name '>' ShiftExpression
/.$putCase consumeBinaryExpressionWithName(OperatorIds.GREATER); $break ./
RelationalExpression_NotName ::= RelationalExpression_NotName '<=' ShiftExpression
/.$putCase consumeBinaryExpression(OperatorIds.LESS_EQUAL); $break ./
RelationalExpression_NotName ::= Name '<=' ShiftExpression
/.$putCase consumeBinaryExpressionWithName(OperatorIds.LESS_EQUAL); $break ./
RelationalExpression_NotName ::= RelationalExpression_NotName '>=' ShiftExpression
/.$putCase consumeBinaryExpression(OperatorIds.GREATER_EQUAL); $break ./
RelationalExpression_NotName ::= Name '>=' ShiftExpression
/.$putCase consumeBinaryExpressionWithName(OperatorIds.GREATER_EQUAL); $break ./
/:$readableName Expression:/

Compare this with say:

ShiftExpression_NotName -> AdditiveExpression_NotName
ShiftExpression_NotName ::= ShiftExpression_NotName '<<' AdditiveExpression
/.$putCase consumeBinaryExpression(OperatorIds.LEFT_SHIFT); $break ./
ShiftExpression_NotName ::= Name '<<' AdditiveExpression
/.$putCase consumeBinaryExpressionWithName(OperatorIds.LEFT_SHIFT); $break ./
ShiftExpression_NotName ::= ShiftExpression_NotName '>>' AdditiveExpression
/.$putCase consumeBinaryExpression(OperatorIds.RIGHT_SHIFT); $break ./
ShiftExpression_NotName ::= Name '>>' AdditiveExpression
/.$putCase consumeBinaryExpressionWithName(OperatorIds.RIGHT_SHIFT); $break ./
ShiftExpression_NotName ::= ShiftExpression_NotName '>>>' AdditiveExpression
/.$putCase consumeBinaryExpression(OperatorIds.UNSIGNED_RIGHT_SHIFT); $break ./
ShiftExpression_NotName ::= Name '>>>' AdditiveExpression
/.$putCase consumeBinaryExpressionWithName(OperatorIds.UNSIGNED_RIGHT_SHIFT); $break ./
/:$readableName Expression:/

or 

AdditiveExpression_NotName -> MultiplicativeExpression_NotName
AdditiveExpression_NotName ::= AdditiveExpression_NotName '+' MultiplicativeExpression
/.$putCase consumeBinaryExpression(OperatorIds.PLUS); $break ./
AdditiveExpression_NotName ::= Name '+' MultiplicativeExpression
/.$putCase consumeBinaryExpressionWithName(OperatorIds.PLUS); $break ./
AdditiveExpression_NotName ::= AdditiveExpression_NotName '-' MultiplicativeExpression
/.$putCase consumeBinaryExpression(OperatorIds.MINUS); $break ./
AdditiveExpression_NotName ::= Name '-' MultiplicativeExpression
/.$putCase consumeBinaryExpressionWithName(OperatorIds.MINUS); $break ./
/:$readableName Expression:/

**Comment by stephan-herrmann** [comment]
As I'm coming from a distance let me start by trying to paraphrase the fundamental reason of all this _NotName rules, correct me if I'm wrong:

The following looks like the root of this:
```
PrimaryNoNewArray ::= PushLPAREN Expression_NotName PushRPAREN
PrimaryNoNewArray ::= PushLPAREN Name PushRPAREN
```

I.e., there is one rule with "Name" and everything else is factored out into "Expression_NotName". Now we need to avoid that anything below Expression_NotName allows to derive a plain Name, as this would be ambiguous with the above.

At first I thought we need to avoid any Name _prefix_ downstream of Expression_NotName, but the following rules tell me otherwise:
```
Expression_NotName -> AssignmentExpression_NotName
AssignmentExpression_NotName -> ConditionalExpression_NotName
ConditionalExpression_NotName ::= Name '?' Expression ':' ConditionalExpression
```

Ergo, _NotName doesn't exclude a Name-_prefix_, but in that case further tokens (the directly next one?) must disambiguate.

For the set of plain rules (without _NotName) everything leads up to Name as one alternative:
```
PostfixExpression -> Primary
PostfixExpression ::= Name
PostfixExpression -> PostIncrementExpression
PostfixExpression -> PostDecrementExpression
```
whereas Name is missing from
```
PostfixExpression_NotName -> Primary
PostfixExpression_NotName -> PostIncrementExpression
PostfixExpression_NotName -> PostDecrementExpression
```

So, if the above is correct, I seem to be in a position to consider further details of this aspect :)

**Comment by stephan-herrmann** [comment]
Now for the question why `ShiftExpression_NotName` is different from other similar rules:

Could the difference relate to the difficulty of treating a sequence of `<` or `>` as one or several tokens? See the rules around TypeArgument1 and TypeArgument2.

**Comment by stephan-herrmann** [comment]
Next level of "design recovery": 

The chain of productions for expressions seems to implement operator precedence, which requires the RHS of any binop to be an expression with higher precedence than the current.

The recursion on any binop LHS encodes the ability to express sequences at the same precedence like "a + b - c".

Here I'm puzzled about how this applies to RelationalExpression, which seems to allow " a < b > c"?? Perhaps the grammar allows this to give better error messages (like cannot compare boolean with int)?

Actually the variant of RelationalExpression_NotName is closer to what I would expect, as for `>` and `<` it rejects the concatenation of several operators. But then why still allow it for `<=` and `=>`?

Additionally I wonder about pairs like this:
```
RelationalExpression_NotName ::= ShiftExpression_NotName '<' ShiftExpression
RelationalExpression_NotName ::= Name '<' ShiftExpression
```

Isn't that the same as
```
RelationalExpression_NotName ::= ShiftExpression '<' ShiftExpression
```
?

A look into git history shows that all this was introduced with Java 5 (then in `java_1_5.g`), with no significant changes since then in the area at question. So maybe overloading of `<` and `>` for generics indeed had a finger in the pie?

**Comment by srikanth-sankaran** [comment]
> I.e., there is one rule with "Name" and everything else is factored out into "Expression_NotName". Now we need to avoid that anything below Expression_NotName allows to derive a plain Name, as this would be ambiguous with the above.
> 
> At first I thought we need to avoid any Name _prefix_ downstream of Expression_NotName, but the following rules tell me otherwise:

It is simpler to think of any Xyz_NotName as rules that will never derive/generate a plain Name ... 

> The chain of productions for expressions seems to implement operator precedence,

Yeah, This is a standard technique to enforce rule reduction to favor some precedence hierarchy

> Could the difference relate to the difficulty of treating a sequence of `<` or `>` as one or several tokens? See the rules around TypeArgument1 and TypeArgument2.

...

> So maybe overloading of `<` and `>` for generics indeed had a finger in the pie?

Plausibly/Perhaps. For the time being I am willing/happy/content to side step this issue.

**Comment by stephan-herrmann** [comment]
in response to [this comment](https://github.com/srikanth-sankaran)

> > [@srikanth-sankaran](https://github.com/srikanth-sankaran)  can we, pretty please, try to exercise some separation of concerns where it is still possible?
> 
> Point taken. I'll open a separate thread in a day or two. Thanks.

Did my scholding sweep this train of thought under the rug, or did you create a ticket for it?

**Comment by stephan-herrmann** [comment]
> The following breaks as is with the PR [#4667](https://github.com/eclipse-jdt/eclipse.jdt.core/pull/4667) although it doesn't have any casts: ...

For the records: it compiles just fine with latest from #4667

**Comment by stephan-herrmann** [comment]
> Additionally I wonder about pairs like this:
> ```
> RelationalExpression_NotName ::= ShiftExpression_NotName '<' ShiftExpression
> RelationalExpression_NotName ::= Name '<' ShiftExpression
> ```
> 
> Isn't that the same as
> ```
> RelationalExpression_NotName ::= ShiftExpression '<' ShiftExpression
> ```
> ?

The latter would break LALR(1)-ness. I didn't check further ...

**Comment by stephan-herrmann** [comment]
After scanning comments also here I am very happy with the state reached.

**Comment by srikanth-sankaran** [comment]
> in response to [this comment](https://github.com/srikanth-sankaran)
> 
> > > [@srikanth-sankaran](https://github.com/srikanth-sankaran)  can we, pretty please, try to exercise some separation of concerns where it is still possible?
> > 
> > 
> > Point taken. I'll open a separate thread in a day or two. Thanks.
> 
> Did my scholding sweep this train of thought under the rug, or did you create a ticket for it?

I didn't, I will in a day or two 😆

**Comment by srikanth-sankaran** [comment]
> > The following breaks as is with the PR [#4667](https://github.com/eclipse-jdt/eclipse.jdt.core/pull/4667) although it doesn't have any casts: ...
> 
> For the records: it compiles just fine with latest from [#4667](https://github.com/eclipse-jdt/eclipse.jdt.core/pull/4667)

Yes as verified by `org.eclipse.jdt.core.tests.compiler.regression.LambdaExpressionsTest.testLambdaInTernary_05()`

**Comment by srikanth-sankaran** [comment]
> The latter would break LALR(1)-ness. I didn't check further ...

Yes, I saw that too - "uninlining" can cause some context to be lost in an LALR(1) parser resulting in conflicts which won't arise in a canonical LR parser.
